File: 0077-combinations/0077-combinations.cpp
class Solution {
public:
    void backtrack(int n, int k, vector<vector<int>>& res, vector<int>& arr, int i) {
        if(arr.size() == k) {
            res.push_back(arr);
            return;
        }
        else {
            for(int j=i; j<=n; j++) {
                arr.push_back(j);
                backtrack(n, k, res, arr, j+1);
                arr.pop_back();
            }
        }
    }
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> arr;
        backtrack(n, k, res, arr, 1);
        return res;
    }
};

File: 0077-combinations/0077-combinations.py
class Solution:
    def backtrack(self, n: int, k: int, res: List[List[int]], arr: List[int], i: int) -> None:
        if len(arr) == k:
            res.append(arr.copy())
            return
        else:
            for j in range(i, n):
                arr.append(j)
                self.backtrack(n, k, res, arr, j + 1)
                arr.pop()
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        arr = []
        self.backtrack(n+1, k, res, arr, 1)
        return res